main_script.py

```diff
@@ -0,0 +1,229 @@
+    for index, row in quote_rewards.iterrows():
+        number = int(row["num_quotes"])
+        if number > 0:
+            total_quotes += number
+            quote_rewards_per_solver[row["solver"]] = number
+
+    for solver in quote_rewards_per_solver:
+        quote_rewards_per_solver[solver] = (
+            quote_rewards_per_solver[solver] * QUOTE_REWARDS_BUDGET / total_quotes
+        )
+
+    # computing consistency rewards
+    participation_data = execute_participation_rewards_helper(
+        str(start_block), str(end_block)
+    )
+    participation_per_solver = {}
+    total_participation_tokens = 0
+    for index, row in participation_data.iterrows():
+        ll = list(row["competition_info"])
+        num_solutions = len(ll)
+        i = num_solutions - 1
+        while i >= num_solutions - 3 and i >= 0:
+            solver = ll[i]["solverAddress"]
+            if solver in participation_per_solver:
+                participation_per_solver[solver] += 1.2
+            else:
+                participation_per_solver[solver] = 1.2
+            i = i - 1
+            total_participation_tokens += 1.2
+        while i >= 0:
+            solver = ll[i]["solverAddress"]
+            if solver in participation_per_solver:
+                participation_per_solver[solver] += 1
+            else:
+                participation_per_solver[solver] = 1
+            i = i - 1
+            total_participation_tokens += 1
+
+    for solver in participation_per_solver:
+        participation_per_solver[solver] = (
+            participation_per_solver[solver]
+            * final_consistency_rewards_budget
+            / total_participation_tokens
+        )
+
+    # putting everything together
+    ovedrafts = {}
+    final_rewards_per_solver = {}
+    for solver in performance_rewards_per_solver:
+        final_rewards_per_solver[solver] = [
+            performance_rewards_per_solver[solver],
+            0,
+            0,
+            performance_rewards_per_solver[solver],
+        ]
+    for solver in quote_rewards_per_solver:
+        if solver in final_rewards_per_solver:
+            final_rewards_per_solver[solver][1] = quote_rewards_per_solver[solver]
+            final_rewards_per_solver[solver][3] += quote_rewards_per_solver[solver]
+        else:
+            final_rewards_per_solver[solver] = [
+                0,
+                quote_rewards_per_solver[solver],
+                0,
+                quote_rewards_per_solver[solver],
+            ]
+    for solver in participation_per_solver:
+        if solver in final_rewards_per_solver:
+            final_rewards_per_solver[solver][2] = participation_per_solver[solver]
+            final_rewards_per_solver[solver][3] += participation_per_solver[solver]
+        else:
+            final_rewards_per_solver[solver] = [
+                0,
+                0,
+                participation_per_solver[solver],
+                participation_per_solver[solver],
+            ]
```

Comment:
If I understand correctly, what you wrote should be equivalent to the following. It uses the somewhat obscure `|` operation to get the union of keys. But then you do not have to copy the key creation stuff for all loops.
```suggestion
    final_rewards_per_solver = {solver: [0, 0, 0, 0] for solver in performance_rewards_per_solver | quote_rewards_per_solver | participation_per_solver}
    for solver in performance_rewards_per_solver:
        final_rewards_per_solver[solver][0] = performance_rewards_per_solver[solver]
        final_rewards_per_solver[solver][3] += performance_rewards_per_solver[solver]
    for solver in quote_rewards_per_solver:
        final_rewards_per_solver[solver][1] = quote_rewards_per_solver[solver]
        final_rewards_per_solver[solver][3] += quote_rewards_per_solver[solver]
    for solver in participation_per_solver:
        final_rewards_per_solver[solver][2] = participation_per_solver[solver]
        final_rewards_per_solver[solver][3] += participation_per_solver[solver]

```

Reply:
Nice!